MAGPI PlatformIO/src/main.cpp

```diff
@@ -0,0 +1,461 @@
+  mpu.setInterruptPinPolarity(true);
+  mpu.setMotionInterrupt(true);
+  */
+
+}
+
+void set_GPS(){
+  Serial8.begin(9600);
+
+  #ifdef DEBUG
+  if(Serial8){
+    Serial.print("GPS succesfully initialized");
+  } else{
+    Serial.print("Serial8 not open");
+  }
+  #endif
+}
+
+void get_data(){
+  #ifdef REAL
+  
+  bool reset = 0;
+  unsigned long timestamp_GPS_package = 0;
+  
+  elapsedMicros since_last_GPS; //must be declared outside loop for access
+  while(Serial8.available()>0){ // check for gps data
+    if(!reset){
+      timestamp_GPS_package = since_last_GPS;//time stamp for first GPS string package is recieved
+      reset = 1;
+    }
+    myGPS.encode(Serial8.read());// encode 
+  }
+
+  if (myGPS.location.isUpdated()){////////////need check, should perform .isUpdated() for all data?
+    lat = myGPS.location.lat();
+    lon = myGPS.location.lng();
+    gps_time = myGPS.time.value(); //Raw HHMMSSCC (u32)
+    gps_alt = myGPS.altitude.meters();
+    heading = myGPS.course.deg();
+    speed = myGPS.speed.mps();
+
+  } else{
+    current_time = since_last_GPS - timestamp_GPS_package; //////////need check//////////// should use gps time + micros OR just treat it as another variable
+  }
+
+  mpu.getEvent(&a, &w, &temp);
+  bmp.performReading(); //write to public attributes temperature and pressure
+  alt_bmp = bmp.readAltitude(SEALEVELPRESSURE_HPA); 
+  #endif
+
+  #ifdef DEBUG_SERIAL_INPUT
+    serial_debug("lat: ");
+    lat = double(Serial.parseFloat());
+    serial_debug("lon: ");
+    lon = double(Serial.parseFloat());
+    serial_debug("GPS Time (HHMMSSCC): ");
+    current_time = Serial.parseInt(); //UNIX time
+    //gps_alt = myGPS.altitude.meters();
+    serial_debug("heading: ");
+    heading = myGPS.course.deg();
+
+    //w.gyro.x = 0.0;
+
+    //speed = myGPS.speed.mps();
+  #endif
+}
+
+void save_data(){
+  //Use Serial for DEBUG [add pause function]
+  myFile = SD.open(filename, FILE_WRITE);
+  myFile.print(current_time);
+  myFile.print(delim);
+
+  //mpu data sensors readings
+  myFile.print(a.acceleration.x); myFile.print(delim);
+  myFile.print(a.acceleration.y); myFile.print(delim);
+  myFile.print(a.acceleration.z); myFile.print(delim);
+  myFile.print(w.gyro.x); myFile.print(delim);
+  myFile.print(w.gyro.y); myFile.print(delim);
+  myFile.print(w.gyro.z); myFile.print(delim);
+
+  //bmp
+  myFile.print(bmp.temperature); myFile.print(delim);
+  myFile.print(bmp.pressure); myFile.print(delim);
+  myFile.print(alt_bmp); myFile.print(delim);
+
+  myFile.print(lat); myFile.print(delim);
+  myFile.print(lon); myFile.println(""); 
+
+  myFile.close(); 
+  
+}
+
+double principal_angle(double angle){
+    if (angle > 180.0){
+      angle += -360.0;
+    } else if (angle < -180.0){
+      angle += 360.0;
+    } else if (angle == 360.0 || angle == -360.0){
+      angle = 0.0;
+    }
+  return angle;
+}
+
+void rotate_motor(double length){
+  int pos = round(turn_per_length * length);
```

Comment:
This will truncate automatically so you don't need the round()

Reply:
I add round here as I know this is more like floor(), but since the resolution on the motor is high, I guess that does not matter much.